### PR TITLE
Added address sanitization feature for translating ML models to LLVM modules

### PR DESCRIPTION
**Pull request by inho9606**
1. Added an option, -iree-llvm-address-sanitizer. Users can turn it on and it injects ASAN to binary of the module.
2. If ASAN is used on compilation, it changes a bit to true in flat buffer to mark it is compiled with ASAN enabled. IREE runtime can read the bit before running the module. And, it will make an error message if runtime does not have ASAN feature when running address sanitized module.
3. To test prototype of address sanitizer, I wrote a simple testcase trying to access out of index using torch_index_select. When I executed the module, it displayed address sanitizer error message.
4. To test #2, I built two IREE, one is with -DIREE_ENABLE_ASAN_ON option, and another one is without that option. And I checked the one without ASAN option showed IREE_STATUS_UNAVAILABLE error when running a module compiled with ASAN Another one with ASAN option gracefully ran the module.
I hope this works..

### CODE REVIEW

**Comment by hanhanW** [comment]
I think dylib_executable.cc was moved to https://github.com/google/iree/blob/main/iree/hal/local/loaders/legacy_library_loader.cc

@benvanik is it a right file to modify?

**Comment by benvanik** [comment]
> I think dylib_executable.cc was moved to [https://github.com/google/iree/blob/main/iree/hal/local/loaders/legacy_library_loader.cc](https://github.com/google/iree/blob/main/iree/hal/local/loaders/legacy_library_loader.cc?rgh-link-date=2021-02-01T14%3A18%3A43Z)
> 
> @benvanik is it a right file to modify?

yep! it'll be replaced with system_library_loader soonish (#3580) but anything you do in there I'll port over so ignore it for now.

File: iree/compiler/Dialect/HAL/Target/LLVM/LLVMTargetOptions.cpp
Comment (benvanik) [inline iree/compiler/Dialect/HAL/Target/LLVM/LLVMTargetOptions.cpp:111]: If you could change this to `iree-llvm-sanitize` as an enum value then we can better match the compiler ASAN flags people are used to and support other sanitizers in the future (`-fsanitize=address` -> `-iree-llvm-sanitize=address`, =memory, etc etc)

File: iree/compiler/Dialect/HAL/Target/LLVM/LLVMIRPasses.cpp
Comment (benvanik) [inline iree/compiler/Dialect/HAL/Target/LLVM/LLVMIRPasses.cpp:31]: unneeded line

File: iree/compiler/Dialect/HAL/Target/LLVM/LLVMAOTTarget.cpp
Comment (benvanik) [inline iree/compiler/Dialect/HAL/Target/LLVM/LLVMAOTTarget.cpp:449]: here too an enum would allow us to add other sanitizer modes

File: build_tools/cmake/iree_copts.cmake
Comment (benvanik) [inline build_tools/cmake/iree_copts.cmake:423]: what's the compile time cost of this? is it required? (if the user is building their application with asan they aren't using this copy of compiler-rt, and conflicting will be bad)

**Review by benvanik — CHANGES_REQUESTED** [review]

File: iree/compiler/Dialect/HAL/Target/LLVM/LLVMTargetOptions.cpp
Comment (inho9606) [reply iree/compiler/Dialect/HAL/Target/LLVM/LLVMTargetOptions.cpp]: Thanks Ben, I updated this part in my new commit. Now address sanitizer is turned on by -iree-llvm-sanitizer=address.
But, for that, I used the member variable, sanitizerKind, representing sanitizer modes in LLVMTargetOption as int type, and used a flag in flatBuffer as ubyte. Is it okay? or can I use other data type for those?

File: iree/compiler/Dialect/HAL/Target/LLVM/LLVMAOTTarget.cpp
Comment (inho9606) [reply iree/compiler/Dialect/HAL/Target/LLVM/LLVMAOTTarget.cpp]: Yes, here my new commit has options.sanitizerKind is int type, but sanitized_kind in flatbuffer schema is ubyte. Can I use sanitizerKind as char type instead of int? I think one byte is enough to represent sanitizer modes..

File: build_tools/cmake/iree_copts.cmake
Comment (inho9606) [reply build_tools/cmake/iree_copts.cmake]: It was trial when I was trying to test it with ASAN libraries.. And forgot to remove it. Sorry and thanks for letting me know.

**Review by hanhanW — CHANGES_REQUESTED** [review]

File: iree/schemas/dylib_executable_def.fbs
Comment (inho9606) [reply iree/schemas/dylib_executable_def.fbs]: Sorry I thought about defining enum in LLVM options. I added it in flat buffer too. But I couldn't find the way to read enumerators to compare it and sanitized_kind value in table. Actually this is part of dylib_executable.cc, which is moved to another file, but just wondered how to do that..

File: iree/compiler/Dialect/HAL/Target/LLVM/LLVMTargetOptions.h
Comment (inho9606) [reply iree/compiler/Dialect/HAL/Target/LLVM/LLVMTargetOptions.h]: I was thinking that it should be converted to other type.. Thanks for letting me know.
And, I changed some design for enum. Moved enum Sanitizer to inside of LLVMTargetOptions and used enum instead of enum class since I thought this sanitizer option would be used only with LLVM Target option.

**Comment by hanhanW** [comment]
As the offline discussion, please try to use the enum type from schema, and rebase to the upstream/main. Thanks!

File: iree/hal/local/loaders/legacy_library_loader.cc
Comment (benvanik) [inline iree/hal/local/loaders/legacy_library_loader.cc:86]: Include `iree/base/target_platform.h` and use the `IREE_SANITIZER_ADDRESS` macro instead of has_feature directly (that macro works across compiler types).

File: iree/hal/local/loaders/legacy_library_loader.cc
Comment (benvanik) [inline iree/hal/local/loaders/legacy_library_loader.cc:84]: If a new sanitizer kind is added this would fail to catch it and continue trying to load the executable. Instead, you can check to see if the sanitizer kind is set and if so require that the sanitizer is available (a switch statement would work) and otherwise return unavailable.

File: iree/hal/local/loaders/legacy_library_loader.cc
Comment (benvanik) [inline iree/hal/local/loaders/legacy_library_loader.cc:88]: to help users who encounter this it's useful to hint at the precise problem and what to do about it, like: `"dynamic library executable is compiled with ASAN support but the host application is not; enable ASAN to load this executable"`

File: iree/schemas/dylib_executable_def.fbs
Comment (benvanik) [inline iree/schemas/dylib_executable_def.fbs:18]: the naming convention in flatbuffers is no k here (I know it's confusing - every language wants its own style and there's like 4 in this project :) - so just `None`/`Address`.

**Review by benvanik — CHANGES_REQUESTED** [review]

**Comment by benvanik** [comment]
Great progress!

**Comment by benvanik** [comment]
and sorry for the delay - when you push new changes you can click the refresh button here to re-request reviews:
image
(otherwise the PR still looks like you haven't addressed feedback so people are unlikely to look again)

File: iree/compiler/Dialect/HAL/Target/LLVM/LLVMTargetOptions.cpp
Comment (stellaraccident) [inline iree/compiler/Dialect/HAL/Target/LLVM/LLVMTargetOptions.cpp:91]: Let's add an else with an error if the sanitize kind is not recognized (I just hit this by guessing the value of "asan" instead of "address").

Comment (inho9606) [reply iree/compiler/Dialect/HAL/Target/LLVM/LLVMTargetOptions.cpp:91]: Thanks, then what can I use to show the error message? Can I just use printf..?

File: iree/compiler/Dialect/HAL/Target/LLVM/LLVMTargetOptions.h
Comment (inho9606) [reply iree/compiler/Dialect/HAL/Target/LLVM/LLVMTargetOptions.h:18]: Then can I use the same style of enum for compile time and schema? Actually I declared Sanitizer enum in LLVMTargetOptions structure as well, but removed it because the schema has the same enum..

File: iree/compiler/Dialect/HAL/Target/LLVM/LLVMTargetOptions.cpp
Comment (stellaraccident) [reply iree/compiler/Dialect/HAL/Target/LLVM/LLVMTargetOptions.cpp:91]: `llvm::errs () <<` in a pinch. It is also possible to set up the command line option machinery to know about the values. There are some examples somewhere (afk right now), but the docs are not too bad: https://llvm.org/docs/CommandLine.html#id12

If you use the built-in support for alternates, it will print the error messages for you, iirc.

File: iree/hal/local/loaders/legacy_library_loader.cc
Comment (inho9606) [reply iree/hal/local/loaders/legacy_library_loader.cc:84]: But I added the SanitizerKind to flatbuffer in LLVMAOTTarget.cpp using
iree_DyLibExecutable_sanitized_kind(builder, options_.sanitizerKind);
And, I thought it can check the value of sanitizerKind with this condition.. Was it different from what you meant?
Thanks

File: iree/compiler/Dialect/HAL/Target/LLVM/LLVMTargetOptions.h
Comment (hanhanW) [reply iree/compiler/Dialect/HAL/Target/LLVM/LLVMTargetOptions.h:18]: @benvanik Do we want only one enum class in TargetOptions.h, and cast the enum to ubyte when generating an executable?
Or we want two enum classes, one in TargetOptions.h and another in flatbuffer schema? And we can switch-case to embed the information?

**Comment by hanhanW** [comment]
> and sorry for the delay - when you push new changes you can click the refresh button here to re-request reviews:
> image
> (otherwise the PR still looks like you haven't addressed feedback so people are unlikely to look again)

@inho9606  To request a fresh review from a reviewer, in the sidebar of the Conversation tab, click the second icon from right. It will show "Re-request review" if you move the curser on the icon.

File: iree/compiler/Dialect/HAL/Target/LLVM/LLVMTargetOptions.h
Comment (hanhanW) [reply iree/compiler/Dialect/HAL/Target/LLVM/LLVMTargetOptions.h:18]: Talked to Ben in chat, @inho9606 I think we should have an enum in the flatbuffer and an enum in target options - then just cast when building the flatbuffer in the compiler. Please also add a comment to state that the order in enum should be the same.

Comment (inho9606) [reply iree/compiler/Dialect/HAL/Target/LLVM/LLVMTargetOptions.h:18]: Ah I see, I'll update it soon! THanks

File: iree/compiler/Dialect/HAL/Target/LLVM/LLVMTargetOptions.h
Comment (hanhanW) [inline iree/compiler/Dialect/HAL/Target/LLVM/LLVMTargetOptions.h:44]: Let's declare the enum outside the struct, e.g., https://github.com/google/iree/blob/87c96f26bca9e14196071235eeff2c3ce575677c/iree/compiler/Dialect/VM/Conversion/TargetOptions.h#L25-L29

Here is cpp files, so we should follow the naming style, e.g., `kNone`, kAddress`, etc.

And please use `enum class`. See https://abseil.io/tips/86

File: iree/hal/local/loaders/legacy_library_loader.cc
Comment (hanhanW) [reply iree/hal/local/loaders/legacy_library_loader.cc:85]: We need a switch statement. You will get a value from `iree_DyLibExecutableDef_sanitized_kind_get`, and we will have different check depending on the value.

If it is None, then nothing to check. If it is `address`, then we have to check ASAN features. If it s `memory`, we have to check MSAN features, etc. In this context, having a switch statement would be good.

File: iree/schemas/dylib_executable_def.fbs
Comment (hanhanW) [inline iree/schemas/dylib_executable_def.fbs:16]: Add a blank line right before the comment. This usually helps readability.

File: iree/schemas/dylib_executable_def.fbs
Comment (hanhanW) [inline iree/schemas/dylib_executable_def.fbs:45]: The indent is removed. Please add them back.

File: iree/compiler/Dialect/HAL/Target/LLVM/LLVMTargetOptions.cpp
Comment (hanhanW) [reply iree/compiler/Dialect/HAL/Target/LLVM/LLVMTargetOptions.cpp:96]: I prefer setting up the command line option. This is more clear to me.

https://llvm.org/docs/CommandLine.html#selecting-an-alternative-from-a-set-of-possibilities

Here is another example in IREE:

https://github.com/google/iree/blob/main/iree/compiler/Dialect/VM/Conversion/TargetOptions.cpp#L34-L41

note that the IREE example is using a list, and you don't need a list for this case.

**Review by hanhanW — CHANGES_REQUESTED** [review]

File: iree/compiler/Dialect/HAL/Target/LLVM/LLVMAOTTarget.cpp
Comment (hanhanW) [inline iree/compiler/Dialect/HAL/Target/LLVM/LLVMAOTTarget.cpp:235]: Let's use switch-case

File: iree/compiler/Dialect/HAL/Target/LLVM/LLVMAOTTarget.cpp
Comment (hanhanW) [inline iree/compiler/Dialect/HAL/Target/LLVM/LLVMAOTTarget.cpp:348]: please use c++ style casts like static_cast<float>(double_value)

https://google.github.io/styleguide/cppguide.html#Casting

File: iree/compiler/Dialect/HAL/Target/LLVM/LLVMIRPasses.cpp
Comment (hanhanW) [inline iree/compiler/Dialect/HAL/Target/LLVM/LLVMIRPasses.cpp:97]: They are only used in the lambda. Let's move them into the lambda so we don't have to capture values.

File: iree/compiler/Dialect/HAL/Target/LLVM/LLVMTargetOptions.h
Comment (hanhanW) [inline iree/compiler/Dialect/HAL/Target/LLVM/LLVMTargetOptions.h:26]:
```suggestion
// Defines kinds of sanitizer.
```

File: iree/compiler/Dialect/HAL/Target/LLVM/LLVMTargetOptions.h
Comment (hanhanW) [inline iree/compiler/Dialect/HAL/Target/LLVM/LLVMTargetOptions.h:27]:
```suggestion
// The order in enum class should be same as one in flat buffer schema.
```

File: iree/hal/local/loaders/legacy_library_loader.cc
Comment (hanhanW) [inline iree/hal/local/loaders/legacy_library_loader.cc:96]: wrap `case` and `if` with braces.

**Review by hanhanW — CHANGES_REQUESTED** [review]

File: iree/hal/local/loaders/legacy_library_loader.cc
Comment (hanhanW) [inline iree/hal/local/loaders/legacy_library_loader.cc:94]: Could you also add braces for the if-statement?

**Review by hanhanW — APPROVED** [review]
Thanks!

File: iree/hal/local/loaders/legacy_library_loader.cc
Comment (asaadaldien) [inline iree/hal/local/loaders/legacy_library_loader.cc:91]: format-nit:
```
            IREE_STATUS_UNAVAILABLE,
            "Dynamic library executable is compiled with ASAN support, but "
            "this host application failed to enable ASAN to load this "
            "executable");
```

**Review by asaadaldien — APPROVED** [review]

**Comment by stellaraccident** [comment]
Hi - is there anything I can help with to get this landed? (it would be super useful in debugging workflows)

**Comment by inho9606** [comment]
> Hi - is there anything I can help with to get this landed? (it would be super useful in debugging workflows)

I'm just waiting for merge. Please let me know if I need something to update more fore merge.
Thanks

**Comment by stellaraccident** [comment]
> I'm just waiting for merge. Please let me know if I need something to update more fore merge.
> Thanks

It's still showing as a couple of nit comments unresolved. Did you have any updates to push or is this final?

**Comment by inho9606** [comment]
> > I'm just waiting for merge. Please let me know if I need something to update more fore merge.
> > Thanks
> 
> It's still showing as a couple of nit comments unresolved. Did you have any updates to push or is this final?
I think this is final, and at least Hanhan approved it I think.

**Comment by hanhanW** [comment]
This needs an approval from @benvanik because he requested changes before.

**Review by benvanik — APPROVED** [review]

**Pull request merged**
